### ISSUE TITLE
High memory requirements when compiling

### ISSUE THREAD

**Issue by haampie** [Dao-AILab/flash-attention#1043]
When compiling pytorch, which vendors flash-attention 2.5.0, the memory requirements are enormous: 5.3GB per process or so.

I noticed the intermediate files generated by nvcc are ridiculously large as well: when compiling flash_bwd_hdim128_bf16_sm80.cu the temporary flash_bwd_hdim128_bf16_sm80.cudafe1.gpu file is 165MB of C code.

It looks like template specialization gone wrong:

```
$ c++filt < tmpxft_0001c398_00000000-6_flash_bwd_hdim128_bf16_sm80.cudafe1.gpu | grep -o '^[^#]*' > out.cc

$ du -sh out.cc  # without comments
135M    out.cc

$ grep -o tuple out.cc | wc -l
2213852
```

millions of references to `cute::tuple`.

Is this intentional? Can you reduce memory requirements?

I've submitted a workaround for pytorch to avoid OOM errors: https://github.com/pytorch/pytorch/pull/130443 but would hope the underlying problem can be fixed itself.

**Comment by tridao** [comment]
Yep i'd love to understand why compile takes so much memory. We do use a lot of templating, but I don't quite get how that translates to a very large intermediate file.

**Comment by tridao** [comment]
Do you have any insight into "template specialization gone wrong"?

**Comment by haampie** [comment]
Dunno. As a data point, with clang it needs "only" 1.8GB:

```
clang++-17 \
  -x cu \
  -c \
  -o /tmp/flash.o \
  /path/to/flash_attn/kernels/flash_bwd_hdim128_bf16_sm80.cu \
  --cuda-gpu-arch=sm_80 \
  -Xclang -fcuda-allow-variadic-functions \
  -O3 \
  -std=c++17 \
  -DFLASHATTENTION_DISABLE_ALIBI \
  -DHAVE_MALLOC_USABLE_SIZE=1 \
  -DHAVE_MMAP=1 \
  -DHAVE_SHM_OPEN=1 \
  -DHAVE_SHM_UNLINK=1 \
  -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS \
  -DONNXIFI_ENABLE_EXT=1 \
  -DONNX_ML=1 \
  -DONNX_NAMESPACE=onnx_torch \
  -DUSE_EXTERNAL_MZCRC \
  -D_FILE_OFFSET_BITS=64 \
  -DLIBCUDACXX_ENABLE_SIMPLIFIED_COMPLEX_OPERATIONS \
  -D_GLIBCXX_USE_CXX11_ABI=1 \
  -DONNX_NAMESPACE=onnx_torch \
  -DCUB_WRAPPED_NAMESPACE=at_cuda_detail \
  -DCUDA_HAS_FP16=1 \
  -D__CUDA_NO_HALF_OPERATORS__ \
  -D__CUDA_NO_HALF_CONVERSIONS__ \
  -D__CUDA_NO_HALF2_OPERATORS__ \
  -D__CUDA_NO_BFLOAT16_CONVERSIONS__ \
  -DNDEBUG \
  -DTORCH_USE_LIBUV \
  -DCAFFE2_USE_GLOO
```

takes about 1.2GB of memory.

Maybe you can reach out to NVIDIA and particularly cutlass folks to debug it?

**Comment by RuRo** [comment]
Hi. I spent some time trying to debug this issue. I wasn't able to definitively track down the offending piece of code, but here are some of my findings that might be helpful for anyone trying to understand this code base.

1) The "primary" culprits seem to be the `Flash_.*kernel_traits` family of templates and functions that accept them as template arguments. For example, grepping around the intermediate C file provided by @haampie shows that `kernel_traits` structs are mentioned 6651 times and the `compute_dq_dk_dv_1colblock` function is mentioned 1632 times.

2) Initially, I **thought** that the code in `aten/src/ATen/native/transformers/cuda/flash_attn` didn't seem to use any of the "overly specialized" `cute` template stuff (`cute::tuple`, `cute::C`) that make the RAM usage go boom. However, I later realized that the reason why `cute::tuple` doesn't appear anywhere in the code is that `flash_attn` uses `Shape<...>` which is an alias for `cute::tuple<...>` (and `Int<...>` which is an alias for `cute::C<...>`).

It seems to me that the problem is that `Flash.*kernel_traits` and `compute_dq_dk_dv_1colblock` are instantiated with somewhere around 100-1000 of different combinations of template parameters. At the same time, `kernel_traits` and `compute_dq_dk_dv_1colblock` define and instantiate somewhere around 10-100 of different `Shape` (and `Shape`-related) template types. In the end, the compiler has to create a template specialization for each unique combination, and we end up with 200k+ lines of `tuple` code.

I am not sure, where to go from here optimization-wise, since it seems that this over-specialization isn't completely spurious and therefore cannot be removed without a much more in depth understanding of the code and the intentions of the original authors.

**Comment by adamjstewart** [comment]
Any progress on this? Would love to reap the benefits of flash attention but also be able to build PyTorch in parallel...

**Comment by cyril23** [comment]
[building VLLM](https://github.com/vllm-project/vllm/issues/14452) inherently requires building flash attention. Currently, the build process consumes over 16.8 GB of memory per job. Therefore, if you plan to run it in parallel with 10 CPU cores active, ensure you have more than 168 GB of memory available.

**Comment by haampie** [comment]
@tridao seems like people are still hitting this issue? Do you have time to look into it?

Other than fixing the "template specialization gone wrong" you could try `-DCMAKE_CUDA_COMPILER=clang++` and have CMake use Clang as the device compiler.

That seems pretty straightforward, last time it seemed only a few C++ issues had to be fixed to make it work.
